File: src/personal_data_warehouse_apple_notes/state.py
from __future__ import annotations

from dataclasses import asdict, dataclass
from datetime import UTC, datetime
from pathlib import Path
import json
import os
import tempfile
from typing import Any

STATE_SCHEMA_VERSION = 1


@dataclass(frozen=True)
class AppleNoteUploadStateEntry:
    note_id: str
    fingerprint: str
    revision_id: str
    title: str
    modified_at: str
    is_deleted: bool
    metadata_uploaded: bool
    html_uploaded: bool
    attachments_uploaded: bool
    last_success_at: str
    last_failure_at: str = ""
    last_error: str = ""
    last_checked_at: str = ""

    @classmethod
    def from_mapping(cls, value: dict[str, Any]) -> AppleNoteUploadStateEntry | None:
        try:
            return cls(
                note_id=str(value.get("note_id", "")),
                fingerprint=str(value.get("fingerprint", "")),
                revision_id=str(value.get("revision_id", "")),
                title=str(value.get("title", "")),
                modified_at=str(value.get("modified_at", "")),
                is_deleted=bool(value.get("is_deleted", False)),
                metadata_uploaded=bool(value.get("metadata_uploaded", False)),
                html_uploaded=bool(value.get("html_uploaded", False)),
                attachments_uploaded=bool(value.get("attachments_uploaded", False)),
                last_success_at=str(value.get("last_success_at", "")),
                last_failure_at=str(value.get("last_failure_at", "")),
                last_error=str(value.get("last_error", "")),
                last_checked_at=str(value.get("last_checked_at", "")),
            )
        except (TypeError, ValueError):
            return None

    @property
    def complete(self) -> bool:
        return self.metadata_uploaded and self.attachments_uploaded


@dataclass
class AppleNotesUploadState:
    account: str
    store_path: str
    entries: dict[str, AppleNoteUploadStateEntry]

    @classmethod
    def empty(cls, *, account: str, store_path: Path | str) -> AppleNotesUploadState:
        return cls(account=account, store_path=str(store_path), entries={})

    @classmethod
    def load(cls, path: Path, *, account: str, store_path: Path | str) -> AppleNotesUploadState:
        try:
            payload = json.loads(path.read_text())
        except (FileNotFoundError, json.JSONDecodeError, OSError):
            return cls.empty(account=account, store_path=store_path)
        if not isinstance(payload, dict) or payload.get("schema_version") != STATE_SCHEMA_VERSION:
            return cls.empty(account=account, store_path=store_path)
        if str(payload.get("account", "")) != account:
            return cls.empty(account=account, store_path=store_path)
        if str(payload.get("store_path", "")) != str(store_path):
            return cls.empty(account=account, store_path=store_path)
        entries: dict[str, AppleNoteUploadStateEntry] = {}
        raw_entries = payload.get("entries", {})
        if isinstance(raw_entries, dict):
            for key, raw_entry in raw_entries.items():
                if isinstance(raw_entry, dict):
                    entry = AppleNoteUploadStateEntry.from_mapping(raw_entry)
                    if entry is not None:
                        entries[str(key)] = entry
        return cls(account=account, store_path=str(store_path), entries=entries)

    def save(self, path: Path) -> None:
        path.parent.mkdir(parents=True, exist_ok=True)
        payload = {
            "schema_version": STATE_SCHEMA_VERSION,
            "account": self.account,
            "store_path": self.store_path,
            "entries": {key: asdict(entry) for key, entry in sorted(self.entries.items())},
        }
        encoded = json.dumps(payload, sort_keys=True, indent=2).encode("utf-8")
        with tempfile.NamedTemporaryFile(dir=path.parent, prefix=f".{path.name}.", delete=False) as file:
            temp_name = file.name
            file.write(encoded)
            file.write(b"\n")
            file.flush()
            os.fsync(file.fileno())
        os.replace(temp_name, path)

    def entry_for(self, note_id: str) -> AppleNoteUploadStateEntry | None:
        return self.entries.get(note_id)

    def mark_success(
        self,
        *,
        note_id: str,
        fingerprint: str,
        revision_id: str,
        title: str,
        modified_at: datetime,
        is_deleted: bool,
        metadata_uploaded: bool,
        html_uploaded: bool,
        attachments_uploaded: bool,
        now: datetime,
    ) -> None:
        timestamp = now.astimezone(UTC).isoformat()
        self.entries[note_id] = AppleNoteUploadStateEntry(
            note_id=note_id,
            fingerprint=fingerprint,
            revision_id=revision_id,
            title=title,
            modified_at=modified_at.astimezone(UTC).isoformat(),
            is_deleted=is_deleted,
            metadata_uploaded=metadata_uploaded,
            html_uploaded=html_uploaded,
            attachments_uploaded=attachments_uploaded,
            last_success_at=timestamp,
            last_checked_at=timestamp,
        )

    def mark_failure(
        self,
        *,
        note_id: str,
        error: str,
        now: datetime,
    ) -> None:
        existing = self.entries.get(note_id)
        timestamp = now.astimezone(UTC).isoformat()
        self.entries[note_id] = AppleNoteUploadStateEntry(
            note_id=note_id,
            fingerprint=existing.fingerprint if existing else "",
            revision_id=existing.revision_id if existing else "",
            title=existing.title if existing else "",
            modified_at=existing.modified_at if existing else "",
            is_deleted=existing.is_deleted if existing else False,
            metadata_uploaded=existing.metadata_uploaded if existing else False,
            html_uploaded=existing.html_uploaded if existing else False,
            attachments_uploaded=existing.attachments_uploaded if existing else False,
            last_success_at=existing.last_success_at if existing else "",
            last_failure_at=timestamp,
            last_error=error,
            last_checked_at=timestamp,
        )


def default_state_file() -> Path:
    return Path.home() / "Library" / "Application Support" / "personal-data-warehouse" / "apple-notes-upload-state.json"

File: app/internal/uploaders/applenotes/state.go
package applenotes

import (
	"encoding/json"
	"os"
	"sort"
	"time"

	"github.com/zachlatta/personal-data-warehouse/app/internal/uploaders/common"
)

const stateSchemaVersion = int64(1)

// StateEntry is one note's recorded upload.
type StateEntry struct {
	NoteID              string `json:"note_id"`
	Fingerprint         string `json:"fingerprint"`
	RevisionID          string `json:"revision_id"`
	Title               string `json:"title"`
	ModifiedAt          string `json:"modified_at"`
	IsDeleted           bool   `json:"is_deleted"`
	MetadataUploaded    bool   `json:"metadata_uploaded"`
	HTMLUploaded        bool   `json:"html_uploaded"`
	AttachmentsUploaded bool   `json:"attachments_uploaded"`
	LastSuccessAt       string `json:"last_success_at"`
	LastFailureAt       string `json:"last_failure_at"`
	LastError           string `json:"last_error"`
	LastCheckedAt       string `json:"last_checked_at"`
}

// Complete reports whether the note's metadata and attachments landed.
func (e StateEntry) Complete() bool { return e.MetadataUploaded && e.AttachmentsUploaded }

// State is the JSON state file the Notes uploader keeps
// (apple-notes-upload-state.json), in the same shape the Python one wrote.
type State struct {
	Account   string
	StorePath string
	Entries   map[string]StateEntry
}

// DefaultStateFile is the state path.
func DefaultStateFile() string {
	return common.DefaultStateFile("apple-notes-upload-state.json")
}

// EmptyState is a fresh state.
func EmptyState(account, storePath string) *State {
	return &State{Account: account, StorePath: storePath, Entries: map[string]StateEntry{}}
}

// LoadState reads the file, returning an empty state when it is missing,
// unreadable, or written for a different account/store.
func LoadState(path, account, storePath string) *State {
	data, err := os.ReadFile(path)
	if err != nil {
		return EmptyState(account, storePath)
	}
	var payload struct {
		SchemaVersion int64                      `json:"schema_version"`
		Account       string                     `json:"account"`
		StorePath     string                     `json:"store_path"`
		Entries       map[string]json.RawMessage `json:"entries"`
	}
	if err := json.Unmarshal(data, &payload); err != nil || payload.SchemaVersion != stateSchemaVersion || payload.Account != account || payload.StorePath != storePath {
		return EmptyState(account, storePath)
	}
	state := EmptyState(account, storePath)
	for key, raw := range payload.Entries {
		var entry StateEntry
		if err := json.Unmarshal(raw, &entry); err == nil {
			state.Entries[key] = entry
		}
	}
	return state
}

// Save writes the state atomically (sort_keys, indent=2, like the Python).
func (s *State) Save(path string) error {
	entries := map[string]any{}
	for key, entry := range s.Entries {
		entries[key] = map[string]any{
			"note_id":              entry.NoteID,
			"fingerprint":          entry.Fingerprint,
			"revision_id":          entry.RevisionID,
			"title":                entry.Title,
			"modified_at":          entry.ModifiedAt,
			"is_deleted":           entry.IsDeleted,
			"metadata_uploaded":    entry.MetadataUploaded,
			"html_uploaded":        entry.HTMLUploaded,
			"attachments_uploaded": entry.AttachmentsUploaded,
			"last_success_at":      entry.LastSuccessAt,
			"last_failure_at":      entry.LastFailureAt,
			"last_error":           entry.LastError,
			"last_checked_at":      entry.LastCheckedAt,
		}
	}
	encoded, err := common.IndentedJSON(map[string]any{
		"schema_version": stateSchemaVersion,
		"account":        s.Account,
		"store_path":     s.StorePath,
		"entries":        entries,
	})
	if err != nil {
		return err
	}
	return common.WriteFileAtomic(path, append(encoded, '\n'), 0o644)
}

// EntryFor returns the note's entry.
func (s *State) EntryFor(noteID string) (StateEntry, bool) {
	entry, ok := s.Entries[noteID]
	return entry, ok
}

// SortedNoteIDs lists the recorded notes.
func (s *State) SortedNoteIDs() []string {
	ids := make([]string, 0, len(s.Entries))
	for id := range s.Entries {
		ids = append(ids, id)
	}
	sort.Strings(ids)
	return ids
}

// MarkSuccess records an uploaded revision.
func (s *State) MarkSuccess(noteID, fingerprint, revisionID, title string, modifiedAt time.Time, isDeleted, metadataUploaded, htmlUploaded, attachmentsUploaded bool, now time.Time) {
	timestamp := common.ISOFormat(now)
	s.Entries[noteID] = StateEntry{
		NoteID: noteID, Fingerprint: fingerprint, RevisionID: revisionID, Title: title,
		ModifiedAt: common.ISOFormat(modifiedAt), IsDeleted: isDeleted, MetadataUploaded: metadataUploaded,
		HTMLUploaded: htmlUploaded, AttachmentsUploaded: attachmentsUploaded, LastSuccessAt: timestamp, LastCheckedAt: timestamp,
	}
}

// MarkFailure records a failed attempt, keeping the last success.
func (s *State) MarkFailure(noteID, errText string, now time.Time) {
	existing := s.Entries[noteID]
	timestamp := common.ISOFormat(now)
	existing.NoteID = noteID
	existing.LastFailureAt = timestamp
	existing.LastError = errText
	existing.LastCheckedAt = timestamp
	s.Entries[noteID] = existing
}
